flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/backpressure/BackPressureRequestCoordinator.java

```diff
@@ -0,0 +1,365 @@
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A coordinator for triggering and collecting back pressure stats
+ * of running tasks.
+ */
+public class BackPressureRequestCoordinator {
+
+	private static final Logger LOG = LoggerFactory.getLogger(BackPressureRequestCoordinator.class);
+
+	private static final int NUM_GHOST_REQUEST_IDS = 10;
+
+	private final Object lock = new Object();
+
+	/** Executor used to run the futures. */
+	private final Executor executor;
+
+	/** Request time out of a triggered back pressure request. */
+	private final Time requestTimeout;
+
+	/** In progress back pressure requests. */
+	@GuardedBy("lock")
+	private final Map<Integer, PendingBackPressureRequest> pendingRequests = new HashMap<>();
+
+	/** A list of recent request IDs to identify late messages vs. invalid ones. */
+	private final ArrayDeque<Integer> recentPendingRequests = new ArrayDeque<>(NUM_GHOST_REQUEST_IDS);
+
+	/** Request ID counter. */
+	@GuardedBy("lock")
+	private int requestIdCounter;
+
+	/** Flag indicating whether the coordinator is still running. */
+	@GuardedBy("lock")
+	private boolean isShutDown;
+
+	/**
+	 * Creates a new coordinator for the cluster.
+	 *
+	 * @param executor Used to execute the futures.
+	 * @param requestTimeout Request time out of a triggered back pressure request.
+	 */
+	public BackPressureRequestCoordinator(
+			Executor executor,
+			long requestTimeout) {
+
+		checkArgument(requestTimeout >= 0L, "Illegal request timeout: " + requestTimeout);
+
+		this.executor = checkNotNull(executor);
+		this.requestTimeout = Time.milliseconds(requestTimeout);
+	}
+
+	/**
+	 * Triggers a task back pressure stats request to all tasks.
+	 *
+	 * @param tasks Tasks to request.
+	 * @return A future of the completed task back pressure stats.
+	 */
+	CompletableFuture<BackPressureStats> triggerBackPressureRequest(ExecutionVertex[] tasks) {
+
+		checkNotNull(tasks, "Tasks to request must not be null.");
+		checkArgument(tasks.length >= 1, "No tasks to request.");
+
+		// Execution IDs of running tasks
+		ExecutionAttemptID[] triggerIds = new ExecutionAttemptID[tasks.length];
+		Execution[] executions = new Execution[tasks.length];
+
+		// Check that all tasks are RUNNING before triggering anything. The
+		// triggering can still fail.
+		for (int i = 0; i < triggerIds.length; i++) {
+			Execution execution = tasks[i].getCurrentExecutionAttempt();
+			if (execution != null && execution.getState() == ExecutionState.RUNNING) {
+				executions[i] = execution;
+				triggerIds[i] = execution.getAttemptId();
+			} else {
+				return FutureUtils.completedExceptionally(new IllegalStateException("Task " + tasks[i]
+					.getTaskNameWithSubtaskIndex() + " is not running."));
+			}
+		}
+
+		synchronized (lock) {
+			if (isShutDown) {
+				return FutureUtils.completedExceptionally(new IllegalStateException("Shut down"));
+			}
+
+			final int requestId = requestIdCounter++;
```

Comment:
I suggest removing `final` in this method or adding final for all the other variables, otherwise it seems not very consistent.  E.g. In above `ExecutionAttemptID[] triggerIds, Execution[] executions` we have not used the `final`.